### PR TITLE
Use environment variables proxy settings

### PR DESCRIPTION
**Pull request by benkropf**
#1959 

*Description of changes:*
Check if common environment variables `HTTPS_PROXY`/`HTTP_PROXY`/`https_proxy`/`http_proxy` are set if neither the configured variables or Java properties are set. Precedence is configured values, then Java system properties, then environment variables.

Also checks for non proxy hosts using `NO_PROXY` and `no_proxy`, again keeping same precedence as above.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java [hunk 1]
```diff
@@ -616,6 +618,48 @@ private String getSystemProperty(String property) {
...
+     * Returns the value for the given environment variable.
+     */
+    private String getEnvironmentVariable(String variable) {
+        return System.getenv(variable);
+    }
+
+    /**
+     * Returns a URL object for the given environment variable, if can be constructed, null otherwise.
+     */
+    private URL getUrlFromEnvironmentVariable(String variable) {
```
Comment (spfink) [inline aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java]: Can we rename variable to environmentVariable? Here and in all new places.

File: aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java [hunk 2]
```diff
@@ -655,18 +699,49 @@ private String getProxyHostProperty() {
...
+        if (getProtocol() == Protocol.HTTPS) {
+            if (getHttpsProxyUrlFromEnvironment() != null) {
+                return getHttpsProxyUrlFromEnvironment().getHost();
+            }
+        } else {
+            if (getHttpProxyUrlFromEnvironment() != null) {
+                return getHttpProxyUrlFromEnvironment().getHost();
+            }
+        }
+        return null;
```
Comment (spfink) [inline aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java]: Can you change the logic here to match the system property logic? For example, not returning null here and catching it higher.

File: aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java (hunk 1 above)
Comment (benkropf) [reply aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java]: Good catch, fixed in the newest commit

File: aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java (hunk 2 above)
Comment (benkropf) [reply aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java]: Good point, I felt like this logic was a little clunky. The complication came from supporting both `HTTP_PROXY` and `http_proxy` so the logic was a little more complicated than the system property logic. The newest commit should clean it up a bit. Let me know what you think

**Comment by benkropf** [comment]
@debora-ito can you take a look at this?

**Comment by benkropf** [comment]
+@dagnir

**Comment by dresnick-sf** [comment]
Could this please be merged to master and released? @spfink

**Comment by sf-mkirilov** [comment]
I think that this PR should include some guidelines about the NO_PROXY format env variable. For example, curl defines it as comma-separated list of hosts while the SdkProxyRoutePlanner expects them to be '|' separated.

**Comment by benkropf** [comment]
@sf-mkirilov good point, right now its expecting a `|` separated list because that was the standard defined by Oracle [here](http://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html). However, I think following the same convention as curl makes more sense for environment variables, since that's what GNU defines [here](http://www.gnu.org/software/wget/manual/html_node/Proxies.html) I'm open to suggestions on this one, but for the time being, I'll change it to be a comma separated list and document that.

**Comment by spfink** [comment]
Merged and will go out with tomorrow's release.

**Pull request merged**